aws/rust-runtime/aws-config/src/imds/client.rs

```diff
@@ -206,23 +206,26 @@ impl Client {
             .call(operation)
             .await
             .map_err(|err| match err {
-                SdkError::ConstructionFailure(err) => match err.downcast::<ImdsError>() {
-                    Ok(token_failure) => *token_failure,
-                    Err(other) => ImdsError::Unexpected(other),
-                },
-                SdkError::TimeoutError(err) => ImdsError::IoError(err),
-                SdkError::DispatchFailure(err) => ImdsError::IoError(err.into()),
-                SdkError::ResponseError { err, .. } => ImdsError::IoError(err),
-                SdkError::ServiceError {
-                    err: InnerImdsError::BadStatus,
-                    raw,
-                } => ImdsError::ErrorResponse {
-                    response: raw.into_parts().0,
+                SdkError::ConstructionFailure(_) if err.source().is_some() => {
+                    match err.into_source().map(|e| e.downcast::<ImdsError>()) {
+                        Ok(Ok(token_failure)) => *token_failure,
+                        Ok(Err(err)) => ImdsError::Unexpected(err),
+                        Err(err) => ImdsError::Unexpected(err.into()),
+                    }
+                }
+                SdkError::ConstructionFailure(_) => ImdsError::Unexpected(err.into()),
+                SdkError::ServiceError(context) => match context.err() {
+                    InnerImdsError::InvalidUtf8 => {
+                        ImdsError::Unexpected("IMDS returned invalid UTF-8".into())
+                    }
+                    _ => ImdsError::ErrorResponse {
+                        response: context.into_raw().into_parts().0,
```

Comment:
I don't think it would crash if more variants are added since there isn't any unwrapping taking place, but I will revise it to use both variants rather than a catch all just so that the area is appropriately updated if another variant is added.